### PR TITLE
Feat/allow int candidates

### PR DESCRIPTION
**Pull request by graceg571**
# Summary

Addresses #359 

This PR allows integer and string candidates as input to rank and score ballots. The user can now input a ranking or score using candidates "1" and 1, and each will be treated as separate candidates with a warning raised to let the user know that's the case.

`RankProfile` and `ScoreProfile` store an internal df of the ballots using integer IDs to represent candidate sets or candidates respectively. Now the values of the dataframe are integers. This provides a uniform internal representation regardless of candidate type which will be helpful for our eventual move to Rust. If the df attribute is accessed, the internal df will be translated to its candidate sets/names and cached for future references.

# Why

Scottish election data labels their candidates as integers and stores a mapping of their integer IDs to candidate names. Likewise, it is common within math literature to use integers to represent candidates. This change will align better with the Scottish election data and math literature as well as allow users to more easily input candidates to ballots without casting all to strings.

We could have cast all integers to strings or enforced no mixing of candidate types, but decided against it: election result candidate names would be misaligned with the ballots the user created, and with the internal df there is no need to constrain candidate types across an entire ballot or preference profile.

# Changes
- `types.py`: added `Candidate : TypeAlias = str | int` and `CandidateFloatDictLike` alias. All type annotations are updated to use `Candidate` across the codebase.
- `ballot.py`: `RankBallot` and `ScoreBallot` now accept `Candidate` in rankings and scores. Warning raised when string and integer candidates collide (e.g. "1" and 1), treated as separate candidates
- `pref_profile.py`: `RankProfile` and `ScoreProfile` maintain and internal `_df` with the candidate integer IDs as the values or column names respectively. The public `df` property translates back to candidate names and is cached. `candidate_id_map` is the mapping of candidates to their integer IDs. `id_candidate_map` is the mapping of candidate integer IDs to candidates. For `RankProfile`, the `candidate_id_map` is the mapping of candidate sets to their IDs. These maps allow translation both ways. This allows the values of the `RankProfile.df` to be integers instead of frozensets of candidates. For `ScoreProfile`, the `candidate_id_map` is the mapping of candidate names to their IDs because the `ScoreProfile.df` has the candidate names as their columns, not sets. A new csv version is introduced. `,v2` is added to the header row of the csv to indicate this new format. `to_csv()` encodes the candidate type within the candidate mapping row tuples e.g. `"(A:str:0), (1:int:1)"`. `from_csv()` reads these tuples, confirms the csv version, and for v2, uses the type to cast the candidate. 
- ballot generators (standard and bloc slate generators), elections, utils, graphs, plots, etc. have been updated to accept `Candidate` types as input and propagate `Candidate` types to internal functions
- `sorting.py` adds `sort_candidates_lexicographically(cand: Iterable[Candidate])->tuple[Candidate]` to enable sorting of mixed type candidates. First, the function will try sorting using `sorted()` if that fails due to mixed types, then the integer candidates will be sorted prior to string candidates. 

# Testing
- `test_ScoreBallot.py`/`test_RankBallot.py`: new tests for int and mixed str/int candidates at the ballot level, including collision warning
- `test_rank_pp.py`/`test_score_pp.py`: new tests for int and mixed str/int candidate profiles
- `test_rank_pp_df.py`/`test_score_pp_df.py`: new tests verifying internal `_df` uses integer IDs and the public `df` translates back to candidate names of mixed types 
- `test_mixed_candidates.py`: fuzz tests running Rank and Score based election systems and all `utils.py` functions over profiles with mixed str/int candidates at varying ballot counts (10, 1000, 10000)
- Added mixed candidate tests to `slate_bloc_generator` modules
- `test_multi_profile_bar_plot_snapshots.py`- added a basic snapshot test with mixed candidates

### CODE REVIEW

File: src/votekit/elections/election_types/ranking/plurality_veto.py [hunk 1]
```diff
@@ -430,22 +440,24 @@ class PluralityVeto(_IterativeVetoBase):
...
+        self, scores: dict[Candidate, float]
+    ) -> tuple[frozenset[Candidate], frozenset[Candidate]]:
         """
         Processes vetoes until some candidate's score reaches zero.
 
         Each voter decrements the score of their least favorite remaining candidate.
 
         Args:
-            scores (dict[str, float]): Mutable score dict, modified in place.
+            scores (dict[str | int, float]): Mutable score dict, modified in place.
```
Comment (peterrrock2) [inline src/votekit/elections/election_types/ranking/plurality_veto.py]: Candidates?

File: src/votekit/graphs/pairwise_comparison_graph.py
Comment (peterrrock2) [inline src/votekit/graphs/pairwise_comparison_graph.py:1]: Test this with a mixture of candidate types

File: src/votekit/graphs/pairwise_comparison_graph.py [hunk 2]
```diff
@@ -191,22 +193,34 @@ def get_dominating_tiers_digraph(graph: nx.DiGraph) -> list[set[str]]:
...
+    cand_subset: list[Candidate]
+    | tuple[Candidate]
+    | set[Candidate]
+    | list[str]
+    | set[str]
+    | list[int]
+    | set[int],
+    pairwise_dict: dict[tuple[Candidate, Candidate], tuple[float, float]]
+    | dict[tuple[str, str], tuple[float, float]]
+    | dict[tuple[int, int], tuple[float, float]],
```
Comment (peterrrock2) [inline src/votekit/graphs/pairwise_comparison_graph.py]: Sequence, maybe?

File: src/votekit/matrices/candidate/comentions.py [hunk 3]
```diff
@@ -1,43 +1,46 @@
...
     Args:
-      i (str): Candidate name.
-      j (str): Candidate name.
+      i (Candidate): Candidate name.
+        Candidates can be strings, integers, or mix of both.
+      j (Candidate): Candidate name.
+        Candidates can be strings, integers, or mix of both.
       ballot (RankBallot): RankBallot.
 
     Returns:
```
Comment (peterrrock2) [inline src/votekit/matrices/candidate/comentions.py]: While we are here, maybe rename these?

File: src/votekit/plots/profiles/multi_profile_bar_plot.py [hunk 4]
```diff
@@ -16,25 +17,29 @@
...
             profiles to plot statistics for.
-        stat_function (Callable[[RankProfile | ScoreProfile], dict[str, float]]): Which stat
+        stat_function (Callable[[RankProfile | ScoreProfile], dict[str | int, float]]): Which stat
             to use for the bar plot. Must be a callable that takes a profile and returns
-            a dict with str keys and float values.
+            a dict with str and int keys and float values.
 
     Returns:
-        dict[str, dict[str, float]]: Data dictionary for ``multi_bar_plot``.
+        dict[str, dict[str | int, float]]: Data dictionary for ``multi_bar_plot``.
```
Comment (peterrrock2) [inline src/votekit/plots/profiles/multi_profile_bar_plot.py]: Candidate?

File: src/votekit/plots/profiles/multi_profile_bar_plot.py [hunk 5]
```diff
@@ -16,25 +17,29 @@
 
 ProfileT = TypeVar("ProfileT", bound=PreferenceProfile)
 PlotLabel = str | int
+CandidateList: TypeAlias = list[Candidate] | list[str] | list[int]
+CandidatePlotLabelMapping: TypeAlias = (
```
Comment (peterrrock2) [inline src/votekit/plots/profiles/multi_profile_bar_plot.py]: Move types into types.py

File: src/votekit/plots/profiles/multi_profile_bar_plot.py [hunk 6]
```diff
@@ -16,25 +17,29 @@
 
 ProfileT = TypeVar("ProfileT", bound=PreferenceProfile)
```
Comment (peterrrock2) [inline src/votekit/plots/profiles/multi_profile_bar_plot.py:18]: Check to see if there is a better alternative to this in Python 3.11+

File: src/votekit/plots/profiles/multi_profile_bar_plot.py
Comment (peterrrock2) [inline src/votekit/plots/profiles/multi_profile_bar_plot.py:1]: Run some tests for plotting. Might need to make some snapshot tests

File: src/votekit/plots/profiles/profile_bar_plot.py [hunk 7]
```diff
@@ -41,26 +44,27 @@ def profile_bar_plot(
 
     Args:
         profile (RankProfile): Profile to plot statistics for.
-        stat_function (Callable[[RankProfile], dict[str, float]]): Which stat
-            to use for the bar plot. Must be a callable that takes a profile and returns
-            a dict with str keys and float values.
+        stat_function (Callable[[RankProfile], dict[str | int, float] | dict[str, float]
+            | dict[int, float]]): Which stat to use for the bar plot.
```
Comment (peterrrock2) [inline src/votekit/plots/profiles/profile_bar_plot.py]: Candidate?

File: src/votekit/plots/profiles/profile_bar_plot.py
Comment (peterrrock2) [inline src/votekit/plots/profiles/profile_bar_plot.py:1]: Candidate across this file

File: src/votekit/plots/bar_plot.py [hunk 8]
```diff
@@ -11,24 +11,36 @@
...
+    | Mapping[str, Mapping[str, float]]
+    | Mapping[str, Mapping[int, float]]
+)
+CategoryLabelList: TypeAlias = list[CategoryLabel] | list[str] | list[int]
+CategoryLabelMapping: TypeAlias = (
+    Mapping[CategoryLabel, CategoryLabel]
+    | Mapping[str, CategoryLabel]
+    | Mapping[int, CategoryLabel]
+)
+
```
Comment (peterrrock2) [inline src/votekit/plots/bar_plot.py]: Types.py maybe?

File: src/votekit/plots/bar_plot.py [hunk 9]
```diff
@@ -11,24 +11,36 @@
...
 
     Args:
-        data (dict[str, dict[str, float]]): Categorical data to be cleaned. The value of each dict
-            should be the frequency of the key which is the category name.
+        data (dict[str, dict[str | int, float]]): Categorical data to be cleaned.
+            The value of each dict should be the frequency of the key which is the category name.
 
     Returns:
-        dict[str, dict[str, float]]: Cleaned data.
+        dict[str, dict[str | int, float]]: Cleaned data.
```
Comment (peterrrock2) [inline src/votekit/plots/bar_plot.py:27]: Type on return

File: src/votekit/plots/bar_plot.py
Comment (peterrrock2) [inline src/votekit/plots/bar_plot.py:1]: Check type annotation and doc string consistency here

File: src/votekit/animations.py [hunk 10]
```diff
@@ -273,9 +278,19 @@ def __init__(
...
+        | set[int]
+        | List[Candidate]
+        | List[str]
+        | List[int]
+        | Literal["winners", "viable", "all"] = "viable",
+        nicknames: Optional[dict[Candidate, str] | dict[str, str] | dict[int, str]] = None,
+        candidate_colors: Optional[
+            Mapping[Candidate, ParsableManimColor]
+            | Mapping[str, ParsableManimColor]
+            | Mapping[int, ParsableManimColor]
```
Comment (peterrrock2) [inline src/votekit/animations.py]: Look at types here

File: src/votekit/ballot.py [hunk 11]
```diff
@@ -1,35 +1,45 @@
...
 
-Ranking: TypeAlias = Optional[tuple[frozenset[str], ...]]
-RankingLike: TypeAlias = Optional[Sequence[str | Iterable[str]]]
+from votekit.types import Candidate
+
+Ranking: TypeAlias = Optional[tuple[frozenset[Candidate], ...]]
+RankingLike: TypeAlias = Optional[Sequence[Candidate | Iterable[Candidate]]]
+ScoresLike: TypeAlias = Optional[
+    Mapping[Candidate, float | int] | Mapping[str, float | int] | Mapping[int, float | int]
+]
```
Comment (peterrrock2) [inline src/votekit/ballot.py]: Types.py?

File: src/votekit/ballot.py [hunk 12]
```diff
@@ -194,22 +208,40 @@ def _convert_ranking_candidates_to_frozenset_strip_whitespace(
...
+            for str_cand in str_cands
+            if str_cand.lstrip("-").isdigit() and int(str_cand) in int_cands
+        }
+        if collisions:
+            warnings.warn(
+                f"Candidates {collisions} appear as both str and int (e.g. '1' and 1)."
+                " These will be treated as separate candidates.",
+                UserWarning,
+            )
+
```
Comment (peterrrock2) [inline src/votekit/ballot.py]: Let's take the convention that integer candidates are non-negative so we don't need to worry about '-1' or '--1'  or '---1'

File: src/votekit/utils.py [hunk 13]
```diff
@@ -431,20 +439,21 @@ def borda_scores(
...
     """
     Break a single set of candidates into multiple sets each with a single candidate according
     to a tiebreak rule. Rule 1: random. Rule 2: first-place votes; break the tie based on
     first-place votes in the profile. Rule 3: borda; break the tie based on Borda points in the
     profile. Rule 4: lex/lexicographic/alph/alphabetical; break the tie alphabetically.
 
     Args:
-        r_set (frozenset[str]): Set of candidates on which to break tie.
+        r_set (frozenset[Candidate]): Set of candidates on which to break tie.
+            Candidates can be strings, integers, or mix of both.
```
Comment (peterrrock2) [inline src/votekit/utils.py]: Better name for this var

File: src/votekit/utils.py [hunk 14]
```diff
@@ -723,7 +749,7 @@ def score_profile_from_ballot_scores(
         profile (ScoreProfile): Profile to score.
 
     Returns:
-        dict[str, float]:
+        dict[str | int, float]:
```
Comment (peterrrock2) [inline src/votekit/utils.py]: Candidate?

File: src/votekit/pref_profile/pref_profile.py [hunk 15]
```diff
@@ -353,20 +371,36 @@ def __init__(
...
             max_ranking_length=self.max_ranking_length,
         )
 
+    @cached_property
+    def df(self) -> pd.DataFrame:
+        """
+        Compute the dataframe as a cached property.
+        The dataframe is internally stored with candidate ids.
+        The dataframe will be translated to original candidate names.
+        """
```
Comment (peterrrock2) [inline src/votekit/pref_profile/pref_profile.py:398]: Docstrings are a part of the Public API. Keep internal details in # NOTE comments.

File: src/votekit/pref_profile/pref_profile.py [hunk 16]
```diff
@@ -966,81 +1062,125 @@ def __init__(
...
-            df=self.df,
+            df=self._df,
         )
 
+    @cached_property
+    def df(self) -> pd.DataFrame:
+        """
+        Compute the dataframe as a cached property.
+        The dataframe is internally stored with candidate ids.
+        The dataframe will be translated to original candidate names.
```
Comment (peterrrock2) [inline src/votekit/pref_profile/pref_profile.py]: Internal vs external

File: tests/elections/test_election_model.py
Comment (peterrrock2) [inline tests/elections/test_election_model.py:1]: Why do we need that extra set? Investigate, please.

File: tests/pref_profile/rank_profile/test_rank_pp.py
Comment (peterrrock2) [inline tests/pref_profile/rank_profile/test_rank_pp.py:1]: Check collisions. Also add checks for instantiating from df.

**Review by peterrrock2 — CHANGES_REQUESTED** [review]
- Consistency note: "Candidate can be a str or int" or "Candidate can strings, integers, or a mix of both"
- Fix roundtrip on CSV with mixed types

File: tests/elections/test_election_model.py
Comment (graceg571) [reply tests/elections/test_election_model.py:1]: `elected` and `eliminated` used to be strings. Because strings are iterables, wrapping them in a frozenset was fine. Integers are not iterables and threw an error accordingly.

File: src/votekit/cleaning/rank_profiles_cleaning.py [hunk 17]
```diff
@@ -441,7 +441,7 @@ def _is_equiv_for_remove_and_condense(
 
 
 def remove_and_condense_rank_profile(
-    removed: Candidate | list[Candidate] | list[str] | list[int],
+    removed: Candidate | CandidateListLike,
```
Comment (graceg571) [inline src/votekit/cleaning/rank_profiles_cleaning.py]: Update CandidateListLike

File: src/votekit/elections/election_types/ranking/schulze.py [hunk 18]
```diff
@@ -9,6 +9,7 @@
     pairwise_dict,
 )
 from votekit.pref_profile import RankProfile
+from votekit.sorting import sort_candidates_lexicographically
```
Comment (graceg571) [inline src/votekit/elections/election_types/ranking/schulze.py]: pseudo lexicographical sort with ints and strs. Update sort_candidates_pseudo_lex and alias with longer name sort_candidates_pseudo_lexicographically

Comment (graceg571) [inline src/votekit/elections/election_types/ranking/schulze.py]: move to utils.py, move imports within utils.py functions (`if TYPE_CHECKING:`)

File: tests/pref_profile/data/rank_profile/test_csv_pp_misformat_header_value_2_invalid_tuples.csv [hunk 19]
```diff
@@ -0,0 +1,34 @@
+VoteKit RankProfile
```
Comment (graceg571) [inline tests/pref_profile/data/rank_profile/test_csv_pp_misformat_header_value_2_invalid_tuples.csv:1]: add ",v2" for parser

File: tests/test_mixed_candidates.py [hunk 20]
```diff
@@ -4,11 +4,20 @@
 
 from votekit import utils
 from votekit.ballot_generator import ic_profile_generator
-from votekit.elections import STV, Borda, Plurality
+from votekit.elections import STV, Borda, Plurality, RankedPairs, SimultaneousVeto
```
Comment (graceg571) [inline tests/test_mixed_candidates.py]: test all election types and ballot generators (not necessarily here)

**Review by graceg571 — COMMENTED** [review]
Summary:
- sorting candidates: update name, update logic and example, move into `utils.py`
- Update `CandidateListLike` and similar names
- add `,v2` to csv for easier parsing
- test all election types and ballot generators with mixed candidates

File: src/votekit/elections/election_types/ranking/schulze.py (hunk 18 above)
Comment (graceg571) [reply src/votekit/elections/election_types/ranking/schulze.py]: Moved to utils.py

Comment (graceg571) [reply src/votekit/elections/election_types/ranking/schulze.py]: updated name to include pseudo and included shorter function name

File: tests/pref_profile/data/rank_profile/test_csv_pp_misformat_header_value_2_invalid_tuples.csv (hunk 19 above)
Comment (graceg571) [reply tests/pref_profile/data/rank_profile/test_csv_pp_misformat_header_value_2_invalid_tuples.csv:1]: added, check for it in validation

File: src/votekit/utils.py (hunk 14 above)
Comment (graceg571) [reply src/votekit/utils.py]: Candidate now

File: src/votekit/plots/profiles/multi_profile_bar_plot.py
Comment (peterrrock2) [reply src/votekit/plots/profiles/multi_profile_bar_plot.py:1]: +1

File: src/votekit/plots/profiles/multi_profile_bar_plot.py (hunk 6 above)
Comment (peterrrock2) [reply src/votekit/plots/profiles/multi_profile_bar_plot.py:18]: +1

File: src/votekit/plots/bar_plot.py (hunk 9 above)
Comment (peterrrock2) [reply src/votekit/plots/bar_plot.py:27]: +1

File: src/votekit/matrices/candidate/comentions.py (hunk 3 above)
Comment (graceg571) [reply src/votekit/matrices/candidate/comentions.py]: Done. Renamed to `cand_a` and `cand_b`.

File: src/votekit/elections/election_types/ranking/plurality_veto.py (hunk 1 above)
Comment (graceg571) [reply src/votekit/elections/election_types/ranking/plurality_veto.py]: Changed to `Candidate`. Added docstring about `Candidate` types.

File: src/votekit/graphs/pairwise_comparison_graph.py (hunk 2 above)
Comment (graceg571) [reply src/votekit/graphs/pairwise_comparison_graph.py]: Changed to Sequence

File: src/votekit/plots/profiles/multi_profile_bar_plot.py (hunk 6 above)
Comment (graceg571) [reply src/votekit/plots/profiles/multi_profile_bar_plot.py:18]: No alternative until Python 3.12

**Review by peterrrock2 — CHANGES_REQUESTED** [review]
There are some lingering comments here. 

I went ahead and started to poke around for bugs, and I found a few. Most of them look like they are coming from the validation logic, so we should think about how to centralize that

[int_cand_bugs.py](https://github.com/user-attachments/files/30129337/int_cand_bugs.py)

File: src/votekit/plots/profiles/profile_bar_plot.py
Comment (graceg571) [reply src/votekit/plots/profiles/profile_bar_plot.py:1]: Updated to Candidate

File: src/votekit/plots/bar_plot.py
Comment (graceg571) [reply src/votekit/plots/bar_plot.py:1]: Updated to Candidate and added relevant docstring for `candidate_ordering` and `candidate_legend`

File: tests/pref_profile/rank_profile/test_rank_pp.py
Comment (graceg571) [reply tests/pref_profile/rank_profile/test_rank_pp.py:1]: Added a test to confirm a warning is thrown if candidates collide. The collision check occurs on `self.candidates` which will include all candidates cast from df.

**Comment by graceg571** [comment]
> There are some lingering comments here.
> 
> I went ahead and started to poke around for bugs, and I found a few. Most of them look like they are coming from the validation logic, so we should think about how to centralize that
> 
> [int_cand_bugs.py](https://github.com/user-attachments/files/30129337/int_cand_bugs.py)

Bugs were fixed in most recent commits.

File: src/votekit/plots/profiles/multi_profile_bar_plot.py (hunk 5 above)
Comment (graceg571) [reply src/votekit/plots/profiles/multi_profile_bar_plot.py]: Moved to types.py.

File: src/votekit/plots/profiles/multi_profile_bar_plot.py (hunk 4 above)
Comment (graceg571) [reply src/votekit/plots/profiles/multi_profile_bar_plot.py]: Updated to Candidate.

File: src/votekit/plots/profiles/multi_profile_bar_plot.py
Comment (graceg571) [reply src/votekit/plots/profiles/multi_profile_bar_plot.py:1]: Added multi_profile_bar_plot.py snapshot test with mixed candidates.

**Review by peterrrock2 — APPROVED** [review]
Looking great! Just one thing left here. 

Add edge case tests for 
- ((Jim) Jimmy Guy:str:1) 
- (Jimmy (Jim):str:1) 

in the CSV readers

**Comment by peterrrock2** [comment]
Follow up to deal with 0 score candidates

File: src/votekit/pref_profile/pref_profile.py (hunk 15 above)
Comment (graceg571) [reply src/votekit/pref_profile/pref_profile.py:398]: Moved internal details to #NOTE comments.

File: src/votekit/utils.py (hunk 13 above)
Comment (graceg571) [reply src/votekit/utils.py]: Renamed to `set_to_tiebreak`

File: src/votekit/utils.py [hunk 21]
```diff
@@ -1035,3 +1038,101 @@ def sort_mixed_cands(cand):
...
+    to the same integer candidate when parsed (e.g. '01' vs. 1) is left alone, since it renders as
+    distinguishable text on a plot. Warning fires once per colliding candidate.
+
+    Args:
+        candidates (Sequence[Candidate]): list of candidates.
+
+    Raises:
+        UserWarning: Candidate collision has occurred between a str and int candidate where
+                when casted to to other candidate's type, they are equivalent. Those candidates
+                will be treated as separate candidates.
```
Comment (peterrrock2) [inline src/votekit/utils.py]: Nit: this technically breaks our docstring conventions laid out in CONTRIBUTING.md:

```python
def foo(arg1: str | None, arg2: int = 3) -> str:
    """
    Brief description (try to stay under 100 characters).

    More details if needed.

    Args:
        arg1 (str | None): Description.
        arg2 (int, optional): Description. Defaults to 3.

    Returns:
        str: Description of the returned value.

    Raises:
        ValueError: Description of the failure mode.
    """
```

File: src/votekit/ballot_generator/bloc_slate_generator/config/collections.py [hunk 22]
```diff
@@ -229,6 +241,10 @@ def __init__(
                     "does not implement the '.items()' method."
                 ) from e
 
+    @property
+    def _parent(self) -> "BlocSlateConfig":
```
Comment (peterrrock2) [inline src/votekit/ballot_generator/bloc_slate_generator/config/collections.py]: We should probably move to importing

```python
from __future__ import annotations
```

at the top of the file to avoid having to use the quoted types. So this would become 

```suggestion
    def _parent(self) -> BlocSlateConfig:
```

This is the default behavior in Python 3.14+, so it's  probably worth the switch now.

File: src/votekit/utils.py [hunk 23]
```diff
@@ -1035,3 +1038,101 @@ def sort_mixed_cands(cand):
...
+            )
+
+
+def _validate_candidate_names(candidates: Iterable[Candidate], source: object, attribute: str):
+    """
+    Ensure the candidates are strings or non-negative integers without reserved characters.
+
+    Args:
+        candidates (Sequence[Candidate]): candidates to validate. Can be candidates cast in
+        df or ballots. Or, the candidates defined at the profile level.
```
Comment (peterrrock2) [inline src/votekit/utils.py]:
```suggestion
        candidates (Sequence[Candidate]): candidates to validate. Can be candidates cast in
            df or ballots. Or, the candidates defined at the profile level.
```

File: src/votekit/utils.py [hunk 24]
```diff
@@ -1035,3 +1038,101 @@ def sort_mixed_cands(cand):
...
+            candidates.
+        ValueError: Candidate cannot be '~'. Its a reserved character.
+        ValueError: Candidate cannot contain ':'. Its a reserved character.
+        UserWarning: Candidate collision has occurred between a str and int candidate where
+            when casted to to other candidate's type, they are equivalent. Those candidates
+            will be treated as separate candidates.
+
+    """
+    candidates = list(candidates)
+    source_type = source.__class__.__name__
```
Comment (peterrrock2) [inline src/votekit/utils.py:1099]: We should sanity check and make sure that candidates is not a string before casting it to a list (unlikely since this is internal, but super annoying to run into while working later).

File: tests/ballot_generators/bloc_slate_generators/config/test_bloc_slate_config.py [hunk 25]
```diff
@@ -525,6 +525,162 @@ def test_error_when_cohesion_df_empty(valid_config):
     assert any("Cohesion mapping must be non-empty." in m for m in msgs)
 
 
+# --- Slate candidates-specific errors --------------------------------------
+
+
+def test_error_for_tilda_candidate_in_slate_to_candidates(valid_config):
```
Comment (peterrrock2) [inline tests/ballot_generators/bloc_slate_generators/config/test_bloc_slate_config.py]:
```suggestion
def test_error_for_tilde_candidate_in_slate_to_candidates(valid_config):
```

Misspelled "tilde"

File: tests/pref_profile/score_profile/test_score_pp.py [hunk 26]
```diff
@@ -167,10 +169,187 @@ def test_negative_int_candidate_in_profile_raises_error():
...
+    with pytest.raises(TypeError, match=r"Boolean candidate\(s\) found in ScoreProfile.candidates"):
         ScoreProfile(
             ballots=[
                 ScoreBallot(scores={1: 1}),
             ],
             candidates=[1, True],
         )
+
+
+def test_tilda_candidate_cast_in_profile_raises_error():
```
Comment (peterrrock2) [inline tests/pref_profile/score_profile/test_score_pp.py]:
```suggestion
def test_tilde_candidate_cast_in_profile_raises_error():
```
Misspelled "tilde"

File: src/votekit/ballot_generator/bloc_slate_generator/config/collections.py (hunk 22 above)
Comment (peterrrock2) [reply src/votekit/ballot_generator/bloc_slate_generator/config/collections.py]: This is a super quick find and replace, so we can add it to this PR.

**Review by peterrrock2 — CHANGES_REQUESTED** [review]
Looking great! A couple more small things

File: src/votekit/utils.py (hunk 21 above)
Comment (graceg571) [reply src/votekit/utils.py]: Kept brief description to one line.

File: src/votekit/ballot_generator/bloc_slate_generator/config/collections.py (hunk 22 above)
Comment (graceg571) [reply src/votekit/ballot_generator/bloc_slate_generator/config/collections.py]: Good call. Added.

File: src/votekit/utils.py (hunk 23 above)
Comment (graceg571) [reply src/votekit/utils.py]: Indented correctly now.

File: src/votekit/utils.py (hunk 24 above)
Comment (graceg571) [reply src/votekit/utils.py:1099]: Added sanity check.

File: tests/ballot_generators/bloc_slate_generators/config/test_bloc_slate_config.py (hunk 25 above)
Comment (graceg571) [reply tests/ballot_generators/bloc_slate_generators/config/test_bloc_slate_config.py]: Fixed!

File: tests/pref_profile/score_profile/test_score_pp.py (hunk 26 above)
Comment (graceg571) [reply tests/pref_profile/score_profile/test_score_pp.py]: Fixed!

**Review by peterrrock2 — APPROVED** [review]
LGTM!
